### ISSUE TITLE
[NEXT-782] dynamic `import()` with node16 .js extensions cannot be resolved

### ISSUE THREAD

**Issue by karlhorky** [vercel/next.js#46678]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: linux
      Arch: x64
      Version: #22 SMP Tue Jan 10 18:39:00 UTC 2023
    Binaries:
      Node: 16.17.0
      npm: 8.15.0
      Yarn: 1.22.19
      pnpm: 7.1.0
    Relevant packages:
      next: 13.2.4-canary.0
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Dynamic imports (next/dynamic), TypeScript

### Link to the code that reproduces this issue

https://codesandbox.io/p/sandbox/sleepy-bird-3msu8y?file=%2Fpages%2Findex.tsx

### To Reproduce

1. Open the reproduction
2. Observe that the `Button.js` dynamic import cannot be resolved:
   ```
   ./pages/index.tsx:7:29
   Module not found: Can't resolve './Button.js'
      5 | // Type 'typeof import("/project/sandbox/node_modules/next/dynamic")'
      6 | // has no call signatures.
   >  7 | const Button = dynamic(() => import("./Button.js"), { ssr: false });
        |                             ^
      8 | 
      9 | export default function Home() {
     10 |   return (

   https://nextjs.org/docs/messages/module-not-found
   ```
   
   ![Screenshot 2023-03-02 at 12 38 25](https://user-images.githubusercontent.com/1935696/222419402-25805df2-d3e7-4c0d-8852-d7f70671f5d0.png)

### Describe the Bug

TypeScript `"module": "node16"` support was added by @loettz in the following PR 🎉

- https://github.com/vercel/next.js/pull/44177

However, this support does not work with dynamic `import()`

### Workaround (potentially buggy)

**Edit:** As noted by @tgrassl in [his comment below](https://github.com/vercel/next.js/issues/46678#issuecomment-1689902569), this may lead to other problems such as hydration errors and FOUC.

Adding the following `webpack` config with `extensionAlias` to the `next.config.js` file (see [Workaround 1 in this other issue](https://github.com/vercel/next.js/issues/41961#:~:text=Workaround%201%3A%20resolve.extensionAlias%20configuration%20option)):

```js
/** @type {import("next").NextConfig} */
const config = {
  reactStrictMode: true,
  webpack: (
    /** @type {import('webpack').Configuration} */
    webpackConfig
  ) => {
    // 💥 Still needed for dynamic import
    /** @type {Exclude<typeof webpackConfig.resolve, undefined>} */ (
      webpackConfig.resolve
    ).extensionAlias = {
      ".js": [".ts", ".tsx", ".js", ".jsx"],
    };

    return webpackConfig;
  },
};

export default config;
```

(This is also already present in the reproduction sandbox, just currently commented out)

cc @fluggo @cseas @kachkaev 

### Expected Behavior

Dynamic `import()` of a fully-specified ESM import path (with `.js` extension) should work

<sub>[NEXT-782](https://linear.app/vercel/issue/NEXT-782/dynamic-import-with-node16-js-extensions-cannot-be-resolved)</sub>

**Comment by tgrassl** [comment]
Fixing the extensionAlias also leads to another issue: https://github.com/vercel/next.js/issues/54436

**Comment by karlhorky** [comment]
Thanks, I updated [the Workaround section above](https://github.com/vercel/next.js/issues/46678#issuecomment-1689902569:~:text=Workaround%20(potentially%20buggy)) with your feedback

**Comment by huozhi** [comment]
dynamic import should import the source code file path, in the updated reproduction it's `Button.tsx`. I tested with latest canary and `import('./Button.tsx')` also works well in codesandbox

**Comment by huozhi** [comment]
If you change the extension to `.js`, like `Button.js` you can still import with `Button.js`. Importing a different extension `.js` when the file extension is `.tsx` would be confusing for bundler file resolving. Sorry to come across this issue late, wasn't sure if the original case was `import('./Button.js')` with `Button.js` file in the filesystem, if that's the case it works now.

**Comment by karlhorky** [comment]
> Importing a different extension `.js` when the file extension is `.tsx` would be confusing for bundler file resolving

Yes, it would be confusing, but it's the way that `tsc`, `esbuild`, `tsx`, `webpack` and similar are built, another link for `webpack`, showing that this confusing behavior is why they built the `extensionAlias` option here:

- https://github.com/webpack/webpack/issues/13252

This comes from the original decision of the TypeScript team to not modify the JavaScript you write.

**Comment by karlhorky** [comment]
Not stale

**Comment by karlhorky** [comment]
Still happens with Next.js 16.0.1:

- https://codesandbox.io/p/devbox/modern-sun-wlxqd3?file=%2Fpages%2Findex.tsx%3A16%2C1&workspaceId=ws_GfAuHrswXyA1DoeSwsjjjz

<img width="1561" height="936" alt="Image" src="https://github.com/user-attachments/assets/832a8ee1-f750-465d-a039-281e1b4bfffe" />

**Comment by marcoshernanz** [comment]
Thanks for the report. This is the same fully specified JavaScript-to-TypeScript extension resolution gap tracked in #82945. The bundler should apply an extension alias when a Node-compatible `.js` specifier points to TypeScript source. Closing as a duplicate so the work in #95426 stays consolidated.

**Comment by karlhorky** [comment]
@marcoshernanz I think #82945 only covers Turbopack. Can you either reopen this or change the issue description and title in #82945 to also cover webpack?
